src/nc_server.c

```diff
@@ -925,3 +932,157 @@ server_pool_deinit(struct array *server_pool)
 
     log_debug(LOG_DEBUG, "deinit %"PRIu32" pools", npool);
 }
+
+static struct msg *
+heartbeat_msg_get(struct conn *conn)
+{
+    struct msg *msg;
+
+    ASSERT(conn->client && !conn->proxy);
+
+    msg = msg_get(conn, true, conn->redis);
+    if (msg == NULL) {
+        conn->err = errno;
+    }
+
+    return msg;
+}
+
+static uint32_t
+set_heartbeat_command(struct mbuf *mbuf, int redis)
+{
+#define HEARTBEAT_MEMCACHE_COMMAND "get twemproxy\r\n"
+#define HEARTBEAT_REDIS_COMMAND "*2\r\n$3\r\nget\r\n$9\r\ntwemproxy\r\n"
+    char *command;
+    uint32_t n;
+
+    command = redis ? HEARTBEAT_REDIS_COMMAND : HEARTBEAT_MEMCACHE_COMMAND;
+    n = (uint32_t)strlen(command);
+
+    memcpy(mbuf->last, command, n);
+    ASSERT((mbuf->last + n) <= mbuf->end);
+
+    return n;
+}
+
+static rstatus_t
+send_heartbeat(struct context *ctx, struct conn *conn, struct server *server)
+{
+    struct mbuf *mbuf;
+    uint32_t n;
+    struct msg *msg;
+    struct server_pool *pool;
+    struct conn *c_conn;
+    rstatus_t status;
+
+    pool = (struct server_pool *)(server->owner);
+
+    c_conn = conn_get(pool, true, conn->redis);
+    if (c_conn == NULL) {
+        return NC_ERROR;
+    }
+
+    msg = heartbeat_msg_get(c_conn);
+    if (msg == NULL) {
+        return NC_ERROR;
+    }
+
+    c_conn->rmsg = msg;
+    mbuf = STAILQ_LAST(&msg->mhdr, mbuf, next);
+    if (mbuf == NULL || mbuf_full(mbuf)) {
+        mbuf = mbuf_get();
+        if (mbuf == NULL) {
+            return NC_ERROR;
+        }
+        mbuf_insert(&msg->mhdr, mbuf);
+        msg->pos = mbuf->pos;
+    }
+    ASSERT(mbuf->end - mbuf->last > 0);
+
+    n = set_heartbeat_command(mbuf, conn->redis);
+    mbuf->last += n;
+    msg->mlen += n;
+
+    msg->swallow = 1;
+    server->fail = FAIL_STATUS_ERR_TRY_HEARTBEAT;
+
+    if (TAILQ_EMPTY(&conn->imsg_q)) {
+        status = event_add_out(ctx->evb, conn);
+        if (status != NC_OK) {
+            return status;
+        }
+    }
+    conn->enqueue_inq(ctx, conn, msg);
+    return NC_OK;
+}
+
+void
+server_restore(struct context *ctx, struct conn *conn)
+{
+    struct server *server;
+
+    server = (struct server *)(conn->owner);
+    ASSERT(server != NULL);
+
+    if (server->fail == FAIL_STATUS_NORMAL) {
+        return;
+    }
+
+    send_heartbeat(ctx, conn, server);
+}
+
+rstatus_t
+server_reconnect(struct context *ctx, struct server *server)
+{
+    rstatus_t status;
+    struct conn *conn;
+
+    conn = server_conn(server);
+    if (conn == NULL) {
+        return NC_ERROR;
+    }
+
+    status = server_connect(ctx, server, conn);
+    if (status == NC_OK) {
+        if (conn->connected) {
+            conn->restore(ctx, conn);
+        }
+    } else {
+        server_close(ctx, conn);
+    }
+
+    return status;
+}
+
+void
+add_failed_server(struct context *ctx, struct server *server)
+{
+    struct server **pserver;
```

Comment:
The below comment can/should be ignored if this code were to be combined with a "failover" pool patch in the future to move failing hosts to a separate pool

------

The implementation or caller of add_failed_server should also check if auto_eject_hosts is true?

server_close() calls server_failure for connection failures such as timeouts

```
void
server_close(struct context *ctx, struct conn *conn)
{
    rstatus_t status;
    struct msg *msg, *nmsg; /* current and next message */
    struct conn *c_conn;    /* peer client connection */

    ASSERT(!conn->client && !conn->proxy);

    server_close_stats(ctx, conn->owner, conn->err, conn->eof,
                       conn->connected);

    conn->connected = false;

    if (conn->sd < 0) {
        server_failure(ctx, conn->owner);
```